twoliter/src/lock.rs

```diff
@@ -27,4 +37,208 @@ pub(crate) struct LockedImage {
+            resolve.push(sdk);
+        }
+        while !resolve.is_empty() {
+            let working_set = resolve.clone();
+            resolve = Vec::new();
+            for image in working_set.iter() {
+                if let Some(version) = known.get(&(image.name.clone(), image.vendor.clone())) {
+                    ensure!(
+                        image.version == *version,
+                        "cannot have multiple versions of the same kit ({0}-{1}@{2} != {0}-{3}@{2}",
+                        image.name,
+                        image.version,
+                        image.vendor,
+                        version
+                    );
+                } else {
+                    let vendor = vendor_table.get(&image.vendor).context(format!(
+                        "no vendor '{}' specified in Twoltier.toml",
+                        image.vendor
+                    ))?;
+                    known.insert(
+                        (image.name.clone(), image.vendor.clone()),
+                        image.version.clone(),
+                    );
+                    let locked_image = LockedImage {
+                        name: image.name.clone(),
+                        version: image.version.clone(),
+                        vendor: image.vendor.clone(),
+                        source: format!("{}/{}:v{}", vendor.registry, image.name, image.version),
+                        is_sdk: image.is_sdk,
+                    };
+                    let kit = Self::find_kit(vendor, &locked_image).await?;
+                    locked.push(locked_image);
+                    resolve.push(kit.sdk);
+                    for dep in kit.kit {
+                        resolve.push(dep);
+                    }
+                }
+            }
+        }
+        let sdk_set = locked
+            .iter()
+            .filter(|x| x.is_sdk)
+            .cloned()
+            .collect::<Vec<_>>();
+        ensure!(
+            sdk_set.len() <= 1,
+            "cannot use multiple sdks (found sdk: {})",
+            sdk_set
+                .iter()
+                .map(|x| format!("{}-{}@{}", x.name, x.version, x.vendor))
+                .collect::<Vec<_>>()
+                .join(", ")
+        );
+        let sdk = sdk_set
+            .first()
+            .context("no sdk was found for use, please specify a sdk in Twoliter.toml")?;
+
+        Ok(Self {
+            schema_version: project.schema_version(),
+            release_version: project.release_version().to_string(),
+            digest: project.digest()?,
+            sdk: sdk.clone(),
+            kit: locked.iter().filter(|x| !x.is_sdk).cloned().collect(),
+        })
+    }
+
+    async fn find_kit(vendor: &Vendor, image: &LockedImage) -> Result<ImageMetadata> {
+        // Now inspect the manifest list
+        let manifest_bytes = docker!(
+            ["manifest", "inspect", image.source.as_str()],
+            "failed to find a kit at uri {}",
+            image.source
```

Comment:
nit: it seems slightly less magical just to do the format here:

```suggestion
            format!("failed to find a kit at uri {}", image.source)
```